modules/koblerBidAdapter.js

```diff
@@ -0,0 +1,226 @@
+          currency: res.cur,
+          width: b.w,
+          height: b.h,
+          creativeId: b.crid,
+          dealId: b.dealid,
+          netRevenue: true,
+          ttl: TIME_TO_LIVE_IN_SECONDS,
+          ad: b.adm,
+          nurl: b.nurl,
+          meta: {
+            advertiserDomains: b.adomain
+          }
+        })
+      })
+    });
+  }
+  return bids;
+};
+
+export const onBidWon = function (bid) {
+  const cpm = bid.cpm || 0;
+  const adServerPrice = utils.deepAccess(bid, 'adserverTargeting.hb_pb') || 0;
+  if (utils.isStr(bid.nurl) && bid.nurl !== '') {
+    const winNotificationUrl = utils.replaceAuctionPrice(bid.nurl, cpm)
+      .replace(/\${AD_SERVER_PRICE}/g, adServerPrice);
+    utils.triggerPixel(winNotificationUrl);
+  }
+};
+
+export const onTimeout = function (timeoutDataArray) {
+  if (utils.isArray(timeoutDataArray)) {
+    timeoutDataArray.forEach(timeoutData => {
+      const query = utils.parseQueryStringParameters({
+        ad_unit_code: timeoutData.adUnitCode,
+        auction_id: timeoutData.auctionId,
+        bid_id: timeoutData.bidId,
+        timeout: timeoutData.timeout,
+        placement_id: utils.deepAccess(timeoutData, 'params.0.placementId'),
+        page_url: window.location.href,
+      });
+      const timeoutNotificationUrl = `${TIMEOUT_NOTIFICATION_ENDPOINT}?${query}`;
+      utils.triggerPixel(timeoutNotificationUrl);
+    });
+  }
+};
+
+function buildOpenRtbBidRequestPayload(validBidRequests, bidderRequest) {
+  const imps = validBidRequests.map(br => buildOpenRtbImpObject(br));
+  const timeout = bidderRequest.timeout || config.getConfig('bidderTimeout') || DEFAULT_TIMEOUT;
+  const pageUrl = (bidderRequest.refererInfo && bidderRequest.refererInfo.referer)
+    ? bidderRequest.refererInfo.referer
+    : window.location.href;
+
+  const request = {
+    id: bidderRequest.auctionId,
+    at: 1, // TODO: is the auction always first price?
+    tmax: timeout,
+    cur: [getCurrency()],
+    imp: imps,
+    device: {
+      devicetype: getDevice(),
+      geo: getGeo(validBidRequests[0])
+    },
+    site: {
+      page: pageUrl,
+    },
+    test: getTest(validBidRequests[0])
+  };
+
+  return JSON.stringify(request);
+}
+
+function getCurrency() {
+  return config.getConfig('currency.adServerCurrency') || DEFAULT_CURRENCY;
+}
+
+function buildOpenRtbImpObject(validBidRequest) {
+  const sizes = getSizes(validBidRequest);
+  const mainSize = sizes[0];
+  const floorInfo = getFloorInfo(validBidRequest, mainSize);
+
+  return {
+    id: validBidRequest.bidId,
+    banner: {
+      format: buildFormatArray(sizes),
+      w: mainSize[0],
+      h: mainSize[1],
+      pos: getPosition(validBidRequest)
+    },
+    tagid: validBidRequest.params.placementId,
+    bidfloor: floorInfo.floor,
+    bidfloorcur: floorInfo.currency,
+    pmp: buildPmpObject(validBidRequest)
+  };
+}
+
+function getDevice() {
+  const ws = utils.getWindowSelf();
+  const ua = ws.navigator.userAgent;
```

Comment:
Is there any alternate to use useragent? we should avoid touching useragent for two reasons. 1. it is inaccurate and kind of reading device information. If we don`t have any other way, it is fine.

Reply:
Unfortunately, there isn't. At least all other adapters use a similar approach. I also expected some better (more standard) mechanism for this so I was surprised that there isn't one.

Reply:
okay